[MO] Add reverse_infer (input rank deducing) for set of operations

pavel-esir opened the pull request:
### Details:

This PR is applied on the top of #8767 changes. Added `reverse_infer` 

**1. for all operations where input rank is just equal to output rank:**
- Convert
- Roll
- Identity
- MVN (and all other sorts of normalization)
- etc.

**2. for operations where input rank (or partial shape) is defined, all image detection operations:**
- DetectionOutput
- ExperimentalDetectronPriorGridGenerator
- Proposal
- PriorBoxOp
- ROIAlign
- etc.

**3. non-trivial operations** 
- Gather
- Transpose
- Elementwise (undefined_input_1_rank can be deduced only if `out_rank > defined_input_2_rank`)
- SwapAxis
- LSTM

**4.** `ConvolutionWithGroupsResolver` fixes - now works with dynamic number of channels as well

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: added only for non-trivial cases
* [x]  Framework operation tests: run e2e locally on MXNet models. And also manually checked on TF models with dynamic ranks (all models where it's possible passed, only one mxnet model RNN_transdurec_multibatch is failed to be reshaped to different seq_len).
* [x]  Transformation tests: done for the only one modified transformation `GroupConvolutionResolver`
* [x] Model Optimizer IR Reader check

Documentation:
* n/a  Supported frameworks operations list
* [x]  Guide on how to convert the **public** model: will be done together for all OOB models
* [x]  User guide update: will be done together for all OOB models

### Tickets:
Ticket: CVS-60964, CVS-62076

pavel-esir commented on model-optimizer/extensions/ops/identity.py:
```diff
@@ -27,6 +28,15 @@ def infer(node):
+        if out_shape is not None:
+            node.in_port(0).data.set_shape(out_shape)
+
+        out_value = node.out_port(0).data.get_value()
+        if out_value is not None:
+            node.in_port(0).data.set_value(out_value)
```
redundant

pavel-esir replied:
in the new solution is outdated

pavel-esir commented on model-optimizer/extensions/ops/parameter.py:
```diff
@@ -56,6 +56,9 @@ def reverse_infer(node: Node):
+            return
+        for i, out_port in node.out_ports().items():
+            out_shape = out_port.data.get_shape()
+            if out_shape is not None:
```
add TODO comment:
Add comparison and choose most defined partial shape (e.g. if it contains not only Rank but also number of channels)

pavel-esir replied:
done

vgavrilo commented:
Maybe `MVN`, not `MNV`?

achetver commented on tools/mo/openvino/tools/mo/ops/parameter.py:
```diff
@@ -56,6 +56,11 @@ def reverse_infer(node: Node):
+            return
+        for i, out_port in node.out_ports().items():
+            out_shape = out_port.data.get_shape()
+            # todo: add comparison and choose most defined partial shape,
```
Do we need to implement this logic in this PR or it is a part of future work?

pavel-esir replied:
I didn't have intention to do it this PR, because we anyway at the moment get fully dynamic shapes with only known rank, like `[-1, -1, -1, -1]`.

But I left this comment for the future if we will reversely propagate also partially defined shapes like `[-1, 3, -1, -1]`.

pavel-esir replied:
> BTW looks good. Do we have some real models to check these changes?

I manually checked on a bunch of models:
face-recognition-resnet34-arcface
mobilefacedet-v1-mxnet
octave-resnext-101-0.25
mobilenet-v2-1.0-224
mobilenet-v2-1.4-224
dense_vnet_abdominal_ct
resnet-50-tf
deepvariant_wgs
ResNext101-64x64-mxnet
and bunch of others
and many others

but didn't automatically run. Today I run locally frameworks.ai.openvino.tests on a full list of e2e models. So far results are green. Will update when it's finished

pavel-esir replied:
Anyway, i implemented this logic in the current PR. I check number of dynamic dimensions by `np.ma.count_masked`. 

If Parameter had several outputs and `reverse_infer` deduced shapes not for all of them this logic will choose the best defined one.

achetver commented:
Maybe it will be better to add routine to avoid a lot of similar code in most of `reverse_infer` functions? I mean this:
```
        if node.in_port(0).data.get_shape() is None:
            node.in_port(0).data.set_shape(undefined_shape_of_rank(2))

        if node.in_port(1).data.get_shape() is None:
            node.in_port(1).data.set_shape(undefined_shape_of_rank(2))

        if node.in_port(2).data.get_shape() is None:
            node.in_port(2).data.set_shape(undefined_shape_of_rank(3))

        if node.in_port(3).data.get_shape() is None:
            node.in_port(3).data.set_shape(undefined_shape_of_rank(2))

        if node.in_port(4).data.get_shape() is None:
            node.in_port(4).data.set_shape(undefined_shape_of_rank(2))
```

I think we can create common function and call it with list of input ports to check and list of shape values to set. What do you think?

achetver commented:
BTW looks good. Do we have some real models to check these changes?

pavel-esir commented:
> BTW looks good. Do we have some real models to check these changes?

I manually checked on a bunch of models:
face-recognition-resnet34-arcface
mobilefacedet-v1-mxnet
octave-resnext-101-0.25
mobilenet-v2-1.0-224
mobilenet-v2-1.4-224
dense_vnet_abdominal_ct
resnet-50-tf
deepvariant_wgs
ResNext101-64x64-mxnet
and bunch of others
and many others

but didn't automatically run. Today I run locally frameworks.ai.openvino.tests on a full list of e2e models. So far results are green. Will update when it's finished

pavel-esir commented on tools/mo/openvino/tools/mo/front/common/partial_infer/eltwise.py:
```diff
@@ -80,6 +81,34 @@ def broadcast_dims(dim1, dim2):
+        defined_in_rank = len(defined_in_shape)
+
+        # in defined_input_shape = [1] and output_shape = [x, 400, 400, 3]
+        # we should not loose these information
```
typo: 'in' -> 'if'
rephrase: 'we should not loose these information' -> 'partial shape information about sizes should be lost'

pavel-esir replied:
fixed

achetver commented on tools/mo/openvino/tools/mo/ops/LSTM.py:
```diff
@@ -68,3 +72,29 @@ def infer(node: Node):
+        other_layer_params_size = (hidden_size * direction + hidden_size + 2) * size
+
+        first_layer_params_size = W_size - (num_layers - 1) * other_layer_params_size
+        # first_layer_params_size = (input_size + rnn_layer.hidden_size + 2) * size
```
This line looks confusing. If it is necessary comment then, please, update them to make more clear. If not - remove it.

pavel-esir replied:
agree, will update the comment

pavel-esir replied:
done

pavel-esir commented:
> Maybe it will be better to add routine to avoid a lot of similar code in most of `reverse_infer` functions? I mean this:
> 
> ```
>         if node.in_port(0).data.get_shape() is None:
>             node.in_port(0).data.set_shape(undefined_shape_of_rank(2))
> 
>         if node.in_port(1).data.get_shape() is None:
>             node.in_port(1).data.set_shape(undefined_shape_of_rank(2))
> 
>         if node.in_port(2).data.get_shape() is None:
>             node.in_port(2).data.set_shape(undefined_shape_of_rank(3))
> 
>         if node.in_port(3).data.get_shape() is None:
>             node.in_port(3).data.set_shape(undefined_shape_of_rank(2))
> 
>         if node.in_port(4).data.get_shape() is None:
>             node.in_port(4).data.set_shape(undefined_shape_of_rank(2))
> ```
> 
> I think we can create common function and call it with list of input ports to check and list of shape values to set. What do you think?

Communicated personally. Decided that this request is rather desired than necessary. Can be done separately

achetver reviewed (approved):
Looks good! Approve it, but don't forget about my comment above :)

mvafin commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -303,3 +303,12 @@ def is_dynamic_slice(s: [slice, int, None]):
                                      s.step is dynamic_dimension)
+
+
+def reverse_bypass_infer(node, ports: List[int] = None):
```
What `ports` function argument is needed for?

pavel-esir replied:
`ports` argument is used to specify manually for which inputs ranks/partial_shape should be deduced. 

E.g. if there is a `Split` op, it has 2 inputs and ranks/partial_shapes should be deduced only `in_port = 0`. `in_port = 1` contains only `num_splits` and should not be changed
https://github.com/openvinotoolkit/openvino/blob/f8136c442b9e5ace226ff531d6726551b80b3b09/tools/mo/openvino/tools/mo/ops/split.py#L232

rkazants commented:
@pavel-esir, did you check NetVlad model with dynamic convolutional weights?

rkazants commented on tools/mo/openvino/tools/mo/back/ConvolutionNormalizer.py:
```diff
@@ -33,15 +33,15 @@ def resolve_convolution_with_group(node: Node, group: int, ir_version: str):
         # TODO rewrite this transformation to generate a shape-computing sub-graph. Ticket 62076
```
do we still need to resolve anything?

pavel-esir replied:
Thanks for noticing this. We needed but now I implemented resolving for all cases not only for cases with dynamic channels but even when weights are dynamic.

rkazants commented on tools/mo/openvino/tools/mo/front/common/partial_infer/eltwise.py:
```diff
@@ -80,6 +81,34 @@ def broadcast_dims(dim1, dim2):
+    if output_shape is not None:
+        out_rank = len(output_shape)
+        deduced_in_shape = undefined_shape_of_rank(out_rank)
+
```
what if `out_rank` equal to zero? can we handle this case?

pavel-esir replied:
yes, if `out_rank == 0` `undefined_shape_of_rank` will return valid masked scalar value

rkazants commented on tools/mo/unit_tests/mo/front/common/partial_infer/eltwise_test.py:
```diff
@@ -105,3 +109,68 @@ def test_eltwise_infer_none_min_max(self):
+    def test_reverse_infer_5(self):
+        self.build_and_test_reverse_inference(inp_shape_1=[4, 1],
+                                              inp_shape_2=None,
+                                              out_shape=[dyn, dyn, 4, 1],
```
please add a case when `inp_shape_1=[4, dyn]`, `out_shape=[1, dyn, dyn, 1]`

pavel-esir replied:
Done https://github.com/pavel-esir/openvino/blob/reverse_infer_extended/tools/mo/unit_tests/mo/front/common/partial_infer/eltwise_test.py#L189-L194

But for clarification: such shapes `inp_shape_1=[4, dyn]`, `out_shape=[1, dyn, dyn, 1]` are incompatible.
For out shape to be `1` both `input_1` and `input_2` must be of size `1`. I added checks that initial are compatible. 
For example shapes `inp_shape_1=[1, 10]`, `out_shape=[1,100]` are not compatible and we should raise Error

rkazants commented on tools/mo/unit_tests/mo/front/common/partial_infer/eltwise_test.py:
```diff
@@ -105,3 +109,68 @@ def test_eltwise_infer_none_min_max(self):
+        self.build_and_test_reverse_inference(inp_shape_1=[4, 1],
+                                              inp_shape_2=None,
+                                              out_shape=[dyn, dyn, 4, 1],
+                                              ref_shape=[dyn, dyn, dyn, dyn])
```
also please add test cases when rank/shape deduction is not possible

pavel-esir replied:
Agree. Added checks if deducing is possible and corresponding unit-tests

https://github.com/pavel-esir/openvino/blob/reverse_infer_extended/tools/mo/unit_tests/mo/front/common/partial_infer/eltwise_test.py#L181-L201

rkazants commented on tools/mo/openvino/tools/mo/front/common/partial_infer/eltwise.py:
```diff
@@ -80,6 +81,34 @@ def broadcast_dims(dim1, dim2):
+        out_rank = len(output_shape)
+        deduced_in_shape = undefined_shape_of_rank(out_rank)
+
+        if input_1_shape is not None and input_2_shape is None and out_rank > len(input_1_shape):
```
also I think we need to check `auto_broadcast` attribute of eltwise node and based on it to have different logic. `pdpd` is not needed

pavel-esir replied:
Agree. Implemented different logic depending from `auto_broadcast `.

rkazants commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -314,3 +314,13 @@ def is_dynamic_slice(s: [slice, int, None]):
+def reverse_bypass_infer(node, in_ports: List[int] = None):
+    output_shape = node.out_port(0).data.get_shape()
+    if output_shape is not None:
+        if in_ports is None:
+            in_ports = node.in_ports().keys()
```
please do require explicit input ports because it is risky to change some input port by mistake.

pavel-esir replied:
Agreed and specified in_ports explicitly for operation which have several inputs.

rkazants commented on tools/mo/openvino/tools/mo/ops/gather.py:
```diff
@@ -100,6 +101,38 @@ def infer(node: Node):
+        batch_dims = node.batch_dims
+        batch_dims = batch_dims + len(indices_shape) if batch_dims < 0 else batch_dims
+        axis = node.in_port(2).data.get_value()
+
```
please check that axis is not None
also, please extract scalar axis.
```
        if isinstance(axis, np.ndarray):
            axis = axis.item()
        assert axis is not None, 'axis input is undefined'
```

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/mvn.py:
```diff
@@ -24,7 +25,8 @@ def __init__(self, graph: Graph, attrs: dict):
             'out_ports_count': 1,
-            'infer': self.infer
+            'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
please explicitly specify 0 input port

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/normalize_l2.py:
```diff
@@ -22,7 +22,8 @@ def __init__(self, graph: Graph, attrs: dict):
             'out_ports_count': 1,
-            'infer': self.infer
+            'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
please explicitly specify 0 port

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/parameter.py:
```diff
@@ -75,6 +76,22 @@ def reverse_infer(node: Node):
+        if shape is not None:
+            return
+
+        # choose dimension with the least number of dynamic dimensions
```
I think we can deduce shape dimensions from different output ports and not only select with minimum number of dynamic dimension.

pavel-esir replied:
Agreed and implemented routine `clarify_partial_shape` to do that
https://github.com/openvinotoolkit/openvino/blob/d0fd7a5aae2d0aedf7a1ba97e5877e460e219506/tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py#L338

rkazants commented on tools/mo/openvino/tools/mo/ops/priorbox_clustered.py:
```diff
@@ -72,3 +74,11 @@ def priorbox_clustered_infer(node: Node):
+
+    @staticmethod
+    def reverse_infer(node):
+        if node.in_port(0).data.get_shape() is None:
+            node.in_port(0).data.set_shape(undefined_shape_of_rank(1))
+
+        if node.in_port(1).data.get_shape() is None:
+            node.in_port(1).data.set_shape(undefined_shape_of_rank(1))
```
you can create helper function for such cases and pass there a list with ranks

pavel-esir replied:
Done. Added helper function `set_input_shapes` to specify input shapes for image detection ops. 

It works not with just ranks but with partial shapes because for a lot of image detection ops we know not just a rank but a partial shape. An example for `ROIPooling` :
```
set_input_shapes(node, undefined_shape_of_rank(4), shape_array([dynamic_dimension_value, 5]))
```

rkazants commented on tools/mo/openvino/tools/mo/ops/reverse_sequence.py:
```diff
@@ -18,6 +19,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'in_ports_count': 2,
             'out_ports_count': 1,
             'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
please explicitly specify 0 port

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/roll.py:
```diff
@@ -20,6 +21,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'op': self.op,
             'version': 'opset7',
             'infer': roll_infer,
+            'reverse_infer': reverse_bypass_infer,
```
please explicitly specify zero input port

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/slice_like.py:
```diff
@@ -38,10 +38,10 @@ def infer(node):
         node.out_port(0).data.set_shape(new_shape)
 
-        if input_value is not None:
+        if input_value is not None and is_fully_defined(shape_like):
```
please explain why this is changed

pavel-esir replied:
If we try to do value inference and `shape_like` had dynamic shape then  we will face error in
https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/openvino/tools/mo/ops/slice_like.py#L52-L53

Note that `new_shape` is calculated from `shape_like` and if `shape_like` has dynamic values then `new_shape` has as well. To make it more clean I modified this condition to check if `new_shape` is not dynamic.

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -129,6 +129,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'type': self.op,
 
             'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
please correct. now you set output shape to input. Please create dedicated function. `reverse_bypass_infer` is not suitable here

pavel-esir replied:
Done. Left more detailed answer in the comment below

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -154,6 +155,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'version': 'opset1',
 
             'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
please correct. now you set output shape to input. Please create dedicated function. reverse_bypass_infer is not suitable here

pavel-esir replied:
Agreed and rewritten. Now `reverse_infer` tries get most defined input shapes. If,
```
out_shape_1 = [dyn, 4, 3], out_shape_2 = [7, dyn, 3], axis = 2 
```
then reverse_infer will return source
```
input shape = [7, 4, 6]
```

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -227,6 +229,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'version': 'opset1',
 
             'infer': self.infer,
+            'reverse_infer': lambda x: reverse_bypass_infer(x, in_ports=[0]),
```
please correct. now you set output shape to input. Please create dedicated function. reverse_bypass_infer is not suitable here

pavel-esir replied:
Agreed and rewritten. Now `reverse_infer` tries get most defined input shapes. If,
```
out_shape_1 = [dyn, 4, 3], out_shape_2 = [7, dyn, 3], axis = 2 
```
then reverse_infer will return source
```
input shape = [7, 4, 6]
```

rkazants commented on tools/mo/openvino/tools/mo/ops/swapaxis.py:
```diff
@@ -30,3 +32,11 @@ def infer(node: Node):
+    @staticmethod
+    def reverse_infer(node: Node):
+        output_shape = node.out_port(0).data.get_shape()
+        if output_shape is not None:
```
please insert input_shape is None. it avoids extra copy if input shape is defined

pavel-esir replied:
done

rkazants reviewed (changes requested):
please address comments.

pavel-esir commented on tools/mo/openvino/tools/mo/ops/lrn.py:
```diff
@@ -24,6 +24,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'version': 'opset1',
 
             'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
specify port

pavel-esir replied:
done

pavel-esir commented on tools/mo/openvino/tools/mo/ops/priorbox.py:
```diff
@@ -30,6 +30,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'in_ports_count': 2,
             'out_ports_count': 1,
             'type_infer': self.type_infer,
+            'reverse_infer': self.reverse_infer,
```
specify port

pavel-esir commented on tools/mo/openvino/tools/mo/ops/Reverse.py:
```diff
@@ -19,6 +19,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'in_ports_count': 2,
             'out_ports_count': 1,
             'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/ops/BatchNormInference.py:
```diff
@@ -19,8 +19,10 @@ def __init__(self, graph: Graph, attrs: dict):
             'out_ports_count': 1,
-            'infer': self.infer
+            'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/front/tf/extractors/fused_bn.py:
```diff
@@ -24,5 +24,6 @@ def tf_fused_bn_extractor(pb):
         'data_type': tf_dtype_extractor(pb.attr["T"].type),
         'eps': pb.attr['epsilon'].f,
         'infer': tf_fused_bn_infer,
+        'reverse_infer': reverse_bypass_infer,
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/front/mxnet/extractors/batchnorm.py:
```diff
@@ -3,13 +3,14 @@
         'type': 'BatchNormalization',
         'eps': attrs.float('eps', 0.001),
         'infer': batch_norm_4_infer,
+        'reverse_infer': reverse_bypass_infer,
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/ops/ConvertLike.py:
```diff
@@ -17,6 +17,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'version': 'opset1',
 
             'infer': copy_shape_infer,
+            'reverse_infer': reverse_bypass_infer,
```
fixed, i didn't add unit-test since `reverse_bypass_infer(node, in_ports=[0])` is quite trivial, it just passes output shape into input port 0

pavel-esir commented on tools/mo/openvino/tools/mo/ops/convolution.py:
```diff
@@ -115,8 +115,16 @@ def infer(node: Node):
+            if len(np.array(node.channel_dims).shape) == 0:  # if channel_dims is scalar
+                num_in_channels = input_shape[node.channel_dims]
+            else:  # if channel_dims is an array
+                num_in_channels = input_shape[node.channel_dims][0]
```
Agree. It looks ugly. I implemented a more clean solution. I replaced these lines with 

```
# since item() unmasks values, result should be masked 
backnum_in_channels = shape_array(input_shape[node.channel_dims].item())
```

The reason this is needed is that when `item()` is called it unmasks dynamic dimensions: if `input_shape = [1, dyn, dyn, dyn]` then `input_shape[node.channel_dims].item()` will be equal to `dynamic_dimension_value == -1000000007` instead of `dynamic_dimension == masked`
https://github.com/openvinotoolkit/openvino/blob/master/tools/mo/openvino/tools/mo/ops/convolution.py#L119

To avoid that I referred to elements with `[0]` instead of `item()`. But now solution with `shape_array(...)` is much more clean.

pavel-esir commented on tools/mo/openvino/tools/mo/ops/concat.py:
```diff
@@ -16,6 +17,7 @@ def __init__(self, graph, attrs: dict):
             'version': 'opset1',
             'axis': 1,
             'infer': concat_infer,
+            'reverse_infer': reverse_bypass_infer,
```
I agree that `reverse_bypass_infer` was not correct choice for `Concat`. I've rewritten `reverse_infer`. But I haven't enough time to write unit-tests. 

Considering that for Concat there are not even proper `infer` unit-tests and considering that `reverse_infer` is quite simple I believe we can skip that for this PR and do it separately

pavel-esir commented on tools/mo/openvino/tools/mo/ops/LSTM.py:
```diff
@@ -68,3 +72,30 @@ def infer(node: Node):
+            initial_cell_state_size = node.in_port(3).data.get_shape()
+            if initial_cell_state_size is not None:
+                batch_size = initial_cell_state_size[1]
+
```
Unfortunately we cannot get it directly from input sizes. I answered in the comment below. Let's discuss that there.

pavel-esir commented on tools/mo/openvino/tools/mo/ops/LSTM.py:
```diff
@@ -68,3 +72,30 @@ def infer(node: Node):
+                batch_size = initial_cell_state_size[1]
+
+        input_shape = shape_array([seq_len, batch_size, input_size])
+        node.in_port(0).data.set_shape(input_shape)
```
Answered above. Let's discuss there.

pavel-esir commented on tools/mo/openvino/tools/mo/ops/LSTM.py:
```diff
@@ -68,3 +72,30 @@ def infer(node: Node):
         rnn_infer(node, [1, 2])
+
+    @staticmethod
+    def reverse_infer(node: Node):
```
>it should be the same for all GRU, LSTM, and RNN with custom cell

Not necessary. The most difficulty here is not in recurrence but in the operation itself. For example for LSTM input feature size can be exactly retrieved from W shape but in practice this weight sometimes (e.g. MXNet) is flattened to 1D and we need extra steps to recover W original shape. And as soon as we know original W we can get input feature size.

Indeed most of the calculations happen here
https://github.com/openvinotoolkit/openvino/blob/c66bee26/tools/mo/openvino/tools/mo/ops/LSTM.py#L93-L96

Adding @itikhono. I discussed with him when was implementing this solution.

I agree that now this `reverse_infer` looks tricky but it's not bug of a solution it's property of LSTM itself, that sometimes it's weights are very tricky, e.g. https://github.com/openvinotoolkit/openvino/blob/c66bee260589fc4f3ac31cb9a4359003458aafd2/tools/mo/openvino/tools/mo/middle/MXNetSplitMultiLayers.py#L55-L90

My solution does not looks obvious but at the moment it works. If we are going to improve it I propose to do that as a separate activity.

pavel-esir replied:
Discussed personally. And clarified this solution and moved it into common RNN.py

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -107,6 +108,43 @@ def infer(node):
+            # to get the original source shape [7, 4, 6]
+            # dimensions of axis = 2 must be summed while
+            # for other dimensions clarify_partial_shape should be called
+            splited_dim_size += shapes[-1][axis]
```
what if `shapes[-1][axis]` has dynamic value?

pavel-esir replied:
If `shapes[-1][axis]` is dynamic then `splited_dim_size` becomes dynamic even if subsequent values will be static. And this is expected behavior, in that case we cannot say exactly the size and we set into dynamic.

Left a comment about that in the code.

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -107,6 +108,43 @@ def infer(node):
+        if len(shapes) == 0:
+            return
+
+        precise_pshape = clarify_partial_shape(shapes)
```
you can't use `clarify_partial_shape` because in case all output static shapes it returns incorrect result and does not sum up dimension values

pavel-esir replied:
Discussed personally. If we mask shape[axis] as dynamic then we can `clarify_partial_shape` for Split

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -212,6 +252,38 @@ def infer(node):
+        assert axis is not None, '{} `axis` is unknown for node {}'.format(node.op, node.soft_get('name', node.id))
+
+        shapes = []
+        splited_dim_size = 0
+        for i in range(node['num_splits']):
+            shape = node.out_port(i).data.get_shape()
+            if shape is not None:
+                shapes.append(shape)
+            else:
+                # if at least one output shape is None/undefined we cannot define
+                # exact value of shape along axis
+                splited_dim_size = dynamic_dimension
+                continue
+
+            # if input is splitted into several outputs, e.g.
+            # out_shape_1 = [dyn, 4, 3], out_shape_2 = [7, dyn, 3], axis = 2
+            # to get the original source shape [7, 4, 6]
+            # dimensions of axis = 2 must be summed while
+            # for other dimensions clarify_partial_shape should be called
+            splited_dim_size += shapes[-1][axis]
+            shapes[-1][axis] = dynamic_dimension
+
+        if len(shapes) == 0:
+            return
+        precise_shape = clarify_partial_shape(shapes)
+        precise_shape[axis] = splited_dim_size
+        node.in_port(0).data.set_shape(precise_shape)
```
a lot of duplicated code

pavel-esir replied:
implemented a singe function without duplication

rkazants commented on tools/mo/openvino/tools/mo/ops/BatchNormInference.py:
```diff
@@ -3,7 +3,7 @@
 
 from openvino.tools.mo.graph.graph import Graph
 from openvino.tools.mo.ops.op import Op
-
+from openvino.tools.mo.front.common.partial_infer.utils import reverse_bypass_infer
```
please sort imports and check other files

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/Cast.py:
```diff
@@ -22,6 +23,7 @@ def __init__(self, graph: Graph, attrs: dict):
             'type': 'Convert',
             'version': 'opset1',
             'infer': self.infer,
+            'reverse_infer': reverse_bypass_infer,
```
please specify explicitly input port

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/non_max_suppression.py:
```diff
@@ -116,3 +117,12 @@ def type_infer(node):
+
+    @staticmethod
+    def reverse_infer(node):
+        in_shape_1 = node.in_port(0).data.get_shape()
+        in_shape_2 = node.in_port(1).data.get_shape()
+        if in_shape_1 is None:
+            node.in_port(0).data.set_shape(undefined_shape_of_rank(3))
+        if in_shape_2 is None:
+            node.in_port(1).data.set_shape(undefined_shape_of_rank(3))
```
please use `set_input_shapes`

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/reverse_sequence.py:
```diff
@@ -3,6 +3,7 @@
 
 from openvino.tools.mo.graph.graph import Graph
 from openvino.tools.mo.ops.op import Op, PermuteAttrs
+from openvino.tools.mo.front.common.partial_infer.utils import reverse_bypass_infer
```
please sort imports

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/ops/roialign.py:
```diff
@@ -3,6 +3,8 @@
 
 from openvino.tools.mo.front.common.layout import get_features_dim, shape_for_layout
 from openvino.tools.mo.front.common.partial_infer.utils import compatible_dims
+from openvino.tools.mo.front.common.partial_infer.utils import dynamic_dimension_value, shape_array, \
+    undefined_shape_of_rank, set_input_shapes
```
please combine these two imports

pavel-esir replied:
Done. Also sorted import for all modified files

rkazants reviewed (commented):
please apply comments

pavel-esir commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -314,3 +314,46 @@ def is_dynamic_slice(s: [slice, int, None]):
                                      s.step is dynamic_dimension)
+
+
+def reverse_bypass_infer(node, in_ports: List[int] = None):
```
Done. The only thing is that `DetectionOutput` has 2 optional inputs. This oblige us to add an `if` into its `reverse_infer`. Other ops are OK with such assert

pavel-esir replied:
done

pavel-esir commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -314,3 +314,46 @@ def is_dynamic_slice(s: [slice, int, None]):
+    output_shape = node.out_port(0).data.get_shape()
+    if output_shape is not None:
+        for port in in_ports:
+            if node.in_port(port).data.get_shape() is None:
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -314,3 +314,46 @@ def is_dynamic_slice(s: [slice, int, None]):
+    if in_ports is None:
+        in_ports = [0]
+
+    output_shape = node.out_port(0).data.get_shape()
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -314,3 +314,46 @@ def is_dynamic_slice(s: [slice, int, None]):
+    return out_shape
+
+
+def set_input_shapes(node, *shapes):
```
Done. The only thing is that `DetectionOutput` has 2 optional inputs. This oblige us to add an `if` into its `reverse_infer`. Other ops are OK with such assert

pavel-esir commented on tools/mo/openvino/tools/mo/front/common/partial_infer/utils.py:
```diff
@@ -314,3 +314,46 @@ def is_dynamic_slice(s: [slice, int, None]):
+    :param shapes:
+    :return:
+    """
+    out_shape = shapes[0]
```
done

pavel-esir commented on tools/mo/openvino/tools/mo/ops/RNN.py:
```diff
@@ -154,3 +170,98 @@ def rnn_infer(node: Node, out_ports=None):
+    out_shape = node.out_port(0).data.get_shape()
+    batch_size = dynamic_dimension
+    seq_len = dynamic_dimension
+    in_port_with_initial_states = 3  # initial hidden size values is framework dependent
```
nit: are

pavel-esir commented:
@rkazants added `reverse_infer` for ONNX, MXNet LSTM, GRU, and simple RNN operations. 

For TF situation is a little bit unclear because there are not example networks with LSTM op and also there are no extractors for LSTM, but only for `BlockLSTM` MO internal op. Left meaningful exception if reverse_infer happens for TF `type: RNNSequence` op.

rkazants commented on tools/mo/openvino/tools/mo/back/ConvolutionNormalizer.py:
```diff
@@ -29,22 +33,69 @@ def resolve_convolution_with_group(node: Node, group: int, ir_version: str):
-        assert is_fully_defined(weights_shape[2:]) and is_fully_defined(I) and \
-               np.prod(weights_shape) == np.prod(new_shape), 'Initial weights shape {}, grouped weights shape {}' \
-                                                             ''.format(weights_shape, new_shape)
+        if is_fully_defined(weights_shape[2:]) and is_fully_defined(I):
```
not clear why these changes are needed. Please give description for this transformation and share sub-graphs before and after

pavel-esir replied:
These changes are needed if GoupedConvolution weights input has dynamic spatial shapes. 
For GroupedConv we need to reshape inputs from `[N, C_IN, Z, Y, X]` into `[N, GROUPS * C_IN, Z, Y, X]` and if `X, Y, Z` values are unknown we need to retrieve them from `ShapeOf`. 

https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/convolution/Convolution_1.md?plain=1#L80-L81

https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/convolution/GroupConvolution_1.md?plain=1#L58-L59

We have convolutions with undefined spatial shapes but we don't have GroupedConvolutions with undefined spatial shapes, but if someone tries to convert such models we should be able to process that correctly. This subgraph is inserted only if  any of `X, Y, Z` is undefined and it prepares shape value `[N, GROUPS * C_IN, Z, Y, X]` with the use of `ShapeOf, StridedSlice, Gather and Concat` ops.

At the moment i was not able to get brief picture of subgraph i have only huge. I will try to simplify picture and share asap

pavel-esir replied:
**Before**
![image](https://user-images.githubusercontent.com/5703039/150401838-2915c082-35dc-4516-a3e6-60062cfc21e2.png)

**After** Reshape has weights layout `[N, GROUPS * C_IN, Z, Y, X]`
![image](https://user-images.githubusercontent.com/5703039/150401675-33ba70e2-05f0-403f-8708-f535cc11fb26.png)

rkazants commented on tools/mo/openvino/tools/mo/front/tf/conv_ext.py:
```diff
@@ -16,9 +16,17 @@ class Conv2DFrontExtractor(FrontExtractorOp):
+
+        def get_num_groups(node):
+            num_groups = node.group if 'group' in node else None
+            if num_groups is None and is_fully_defined(node.in_node(0).shape) and is_fully_defined(node.kernel_shape):
+                num_groups = node.in_node(0).shape[node.channel_dims] // node.kernel_shape[node.input_feature_channel]
```
I think it sufficient to check that concrete dimensions are not dynamic: `node.in_node(0).shape[node.channel_dims]` and `node.kernel_shape[node.input_feature_channel]`

pavel-esir replied:
agreed, and corrected these lines as we discussed

rkazants commented on tools/mo/openvino/tools/mo/front/tf/conv_ext.py:
```diff
@@ -16,9 +16,17 @@ class Conv2DFrontExtractor(FrontExtractorOp):
+            num_groups = node.group if 'group' in node else None
+            if num_groups is None and is_fully_defined(node.in_node(0).shape) and is_fully_defined(node.kernel_shape):
+                num_groups = node.in_node(0).shape[node.channel_dims] // node.kernel_shape[node.input_feature_channel]
+            else:
+                num_groups = 1
```
it is not correct. Looks a gap

pavel-esir replied:
Thanks for noticing wrongness of this else. As we discussed I corrected implementation. Now we check not `is_fully_dynamic` but only channel dimensions. Also clarified this `if-elif` statement.

Firstly 
- we try to get number of groups from attribute 
- If attribute is not set  check whether channel dimensions are static and calculate from channel sizes
- and finally if channel sizes are not defined/dynamic set into default num_groups=1

Added corresponding comment into code

rkazants reviewed (commented):
Please fix last comments. The remaining code looks good

pavel-esir commented on tools/mo/openvino/tools/mo/ops/transpose.py:
```diff
@@ -49,4 +51,16 @@ def infer(node):
 
+    @staticmethod
+    def reverse_infer(node: Node):
+        input_shape = node.in_port(0).data.get_shape()
```
I actually do reverse_infer only if input_shape is None.
https://github.com/openvinotoolkit/openvino/blob/6ab34578feea9af43df825a638fdcd9da24fb22e/tools/mo/openvino/tools/mo/ops/transpose.py#L56-L61

I prefer to keep in this Transpose case one `if` instead of having multiple  `if`'s and `returns`. 
This looks shorter:
```
input_shape = node.in_port(0).data.get_shape()
output_shape = node.out_port(0).data.get_shape()
order = node.in_port(1).data.get_value()

if input_shape is None and output_shape is not None and order is not None:
    ...
```
than this:
```
if input_shape is None and output_shape is not None and order is not None:
    ...
input_shape = node.in_port(0).data.get_shape()
if input_shape is None:
    return
output_shape = node.out_port(0).data.get_shape()
if input_shape is None:
    return
order = node.in_port(1).data.get_value()
if input_shape is None:
    return
```

pavel-esir commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -212,6 +228,14 @@ def infer(node):
 
+    @staticmethod
+    def reverse_infer(node: Node):
+        assert hasattr(node, 'num_splits')
```
done

rkazants commented on tools/mo/openvino/tools/mo/ops/split.py:
```diff
@@ -212,6 +227,16 @@ def infer(node):
 
+    @staticmethod
+    def reverse_infer(node: Node):
+        if node.in_port(0).data.get_value() is not None:
```
`get_shape()`?

pavel-esir replied:
thanks for noticing. fixed

rkazants commented on tools/mo/openvino/tools/mo/front/common/partial_infer/eltwise.py:
```diff
@@ -80,8 +82,83 @@ def broadcast_dims(dim1, dim2):
+                    deduced_in_shape[i] = output_shape[i]
+            deduced_in_shape[:-defined_in_rank] = output_shape[:-defined_in_rank]
+
+            node.in_port(in_port_to_update).data.set_shape(deduced_in_shape)
```
please double check. It looks unfixed still.

rkazants commented on tools/mo/openvino/tools/mo/back/ConvolutionNormalizer.py:
```diff
@@ -29,22 +33,69 @@ def resolve_convolution_with_group(node: Node, group: int, ir_version: str):
+        zeros_to_copy_spatial_dims = np.zeros(num_spatial_dims)
+        new_shape = shape_array([node.output, -1, *zeros_to_copy_spatial_dims])
+        reshape = create_op_node_with_second_input(node.graph, Reshape, new_shape,
+                                                   {'override_output_shape': True})
```
please set `special_zero` that is mandatory attribute by the spec

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/back/ConvolutionNormalizer.py:
```diff
@@ -29,22 +33,69 @@ def resolve_convolution_with_group(node: Node, group: int, ir_version: str):
+            assert np.prod(weights_shape) == np.prod(new_shape), 'Initial weights shape {}, grouped weights shape {}' \
+                                                                 ''.format(weights_shape, new_shape)
+            reshape = create_op_node_with_second_input(node.graph, Reshape, new_shape,
+                                                       {'override_output_shape': True})
```
please set special_zero that is mandatory attribute by the spec

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/back/ConvolutionNormalizer.py:
```diff
@@ -29,22 +33,69 @@ def resolve_convolution_with_group(node: Node, group: int, ir_version: str):
+            new_shape_node.in_port(0).connect(const_part_of_shape.out_port(0))
+            new_shape_node.in_port(1).connect(C_IN.out_port(0))
+            new_shape_node.in_port(2).connect(weights_spatial_shape.out_port(0))
+            reshape = Reshape(node.graph, {'override_output_shape': True}).create_node()
```
please set special_zero that is mandatory attribute by the spec

pavel-esir replied:
done

rkazants commented on tools/mo/openvino/tools/mo/back/ConvolutionNormalizer.py:
```diff
@@ -29,22 +33,69 @@ def resolve_convolution_with_group(node: Node, group: int, ir_version: str):
+            weights_shape = Shape(node.graph, {'name': node_name + '/ShapeOfWeights'}).create_node()
+            weights_shape.in_port(0).connect(weights_node.out_port(0))
+
+            weights_spatial_shape = create_op_with_const_inputs(node.graph, StridedSlice,
+                                                                port_value_dict={1: int64_array([2]),
+                                                                                 2: int64_array([-1])},
+                                                                op_attrs={'begin_mask': [1],
+                                                                          'end_mask': [0],
+                                                                          'new_axis_mask': [0],
+                                                                          'shrink_axis_mask': [0],
+                                                                          'ellipsis_mask': [0]},
+                                                                input_node=weights_shape)
```
nit: you can use `node_to_get_shape_value_of_indices()`. Please see an example in `add_reshape_around_convolution.py` from front/kaldi

pavel-esir commented on tools/mo/openvino/tools/mo/ops/slice_like.py:
```diff
@@ -29,7 +29,7 @@ def infer(node):
         shape_like = node.in_port(1).data.get_shape()
 
-        new_shape = np.copy(input_shape)
+        new_shape = shape_array(input_shape.copy())
```
Reverted back into ` shape_array(input_shape.copy())` because if `input_shape` is static then `new_shape` will be a simple numpy array, and if `shape_like` is dynamic then we will try to write `dynamic_dimension` into it and end up with a error

https://github.com/openvinotoolkit/openvino/blob/6ab34578feea9af43df825a638fdcd9da24fb22e/tools/mo/openvino/tools/mo/ops/slice_like.py#L36

![image](https://user-images.githubusercontent.com/5703039/150410924-f72a261a-be68-4bb8-a26c-d4f82a6853bc.png)
This error was actually happening in real model for OOB scenario.

To prevent that I decorate with `shape_array` so that `new_shape` is always a masked numpy type that potentially can store dynamic values.

rkazants commented:
Anticipate DLB results and will merge if no regressions. @pavel-esir, please make pre-commit green

pavel-esir commented:
>First comment, about onnx proposal_ext.py.
you are right, Proposal is not in the onnx public opset. On year 2020 Evegeny Lazarev told me that this extractor is for some custom onnx models trained by internal intel developers.

>Priorbox, priorbox clustered. 
removed `revere_infer` for them

>About Parameter several out_ports.
Thanks for noticing this. Yes, indeed Parameter has only one out_port. Removed extra code.

@jane-intel please take a look

rkazants commented:
@pavel-esir, please fix a test:
```
    def test_clarify_1(self):
>       actual_result = clarify_partial_shape(shape_array([dynamic_dimension, 10, dynamic_dimension]),
                                              shape_array([4, dynamic_dimension, dynamic_dimension]))
E       TypeError: clarify_partial_shape() takes 1 positional argument but 2 were given

../_w/install_pkg/tests/mo/unit_tests/mo/front/common/partial_infer/utils_test.py:152: TypeError
=============================== warnings summary ===============================
../../../../home/AzDevOps/.local/lib/python3.8/site-packages/defusedxml/__init__.py:30
  /home/AzDevOps/.local/lib/python3.8/site-packages/defusedxml/__init__.py:30: DeprecationWarning: defusedxml.cElementTree is deprecated, import from defusedxml.ElementTree instead.
    from . import cElementTree
```

pavel-esir commented:
> @pavel-esir, please fix a test:
> 
> ```
>     def test_clarify_1(self):
> >       actual_result = clarify_partial_shape(shape_array([dynamic_dimension, 10, dynamic_dimension]),
>                                               shape_array([4, dynamic_dimension, dynamic_dimension]))
> E       TypeError: clarify_partial_shape() takes 1 positional argument but 2 were given
> 
> ../_w/install_pkg/tests/mo/unit_tests/mo/front/common/partial_infer/utils_test.py:152: TypeError
> =============================== warnings summary ===============================
> ../../../../home/AzDevOps/.local/lib/python3.8/site-packages/defusedxml/__init__.py:30
>   /home/AzDevOps/.local/lib/python3.8/site-packages/defusedxml/__init__.py:30: DeprecationWarning: defusedxml.cElementTree is deprecated, import from defusedxml.ElementTree instead.
>     from . import cElementTree
> ```

@rkazants done.
